kedro-org/kedro-viz
# Fix: Viz not working with proxy URLs

## Description

Fixes #916

## Development notes

If a user runs Viz with a proxied URL, like you may try to do as described in [this issue](https://github.com/kedro-org/kedro-viz/issues/916), the application would load without critical information, and sometimes no information at all. The reason being is that the URL wouldn't pass our [hard-coded route match](https://github.com/kedro-org/kedro-viz/blob/main/src/components/wrapper/wrapper.js#L60).

In this PR, we'll now allow a dynamic pathname to be recognized and matched, thus allowing the flowchart or experiment-tracking pages to render correctly.

Lastly, we'll also sanitize the URL and remove any duplicative `experiment-tracking` strings from appearing in the pathname. We need to guard against this because, again, if a user has a proxy URL or a base URL that isn't `/`, there will be conflicts. The use case: a user runs Viz with [jupyter-sever-proxy](https://jupyter-server-proxy.readthedocs.io/en/latest/index.html), which adds `/kedro_viz/` to the pathname. So Viz loads, then we navigate to the experiment-tracking page, which does load correctly. Yet if the page is refreshed, we'd receive an incorrect result. If we didn't remove `experiment-tracking` from the pathname, we'd get something like `/kedro_viz/experiment-trackingexperiment-tracking`, and that's not desirable.

## QA notes

Run Viz via `jupyter-sever-proxy` by doing the following:

- Checkout the branch
- `cd` in the `demo-project` folder
- `pip install jupyter-sever-proxy`
- Run `kedro viz`
- Click the `Kedro Viz` icon that appears

All functionality within Viz should function as it should.

## Checklist



<a href="https://gitpod.io/#https://github.com/kedro-org/kedro-viz/pull/941"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
